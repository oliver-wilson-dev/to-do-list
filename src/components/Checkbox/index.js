import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';

const Checkbox = ({ markComplete, id, complete }) => {
  const onChange = () => markComplete({ id });

  return (
    <input
      type="checkbox"
      className={styles.checkbox}
      checked={complete}
      onChange={onChange}
      aria-label={`mark this task as ${complete ? 'incomplete' : 'complete'}`}
    />
  );
};

Checkbox.propTypes = {
  markComplete: PropTypes.func.isRequired,
  id: PropTypes.symbol.isRequired,
  complete: PropTypes.bool.isRequired
};

export default Checkbox;
