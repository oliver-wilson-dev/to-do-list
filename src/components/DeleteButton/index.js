import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';

const DeleteButton = ({ removeToDo, id }) => (
  <button className={styles.button} onClick={removeToDo.bind(undefined, { id })} type="button">
    <span role="img" aria-label="remove this todo">‚ùå</span>
  </button>
);

DeleteButton.propTypes = {
  removeToDo: PropTypes.func.isRequired,
  id: PropTypes.symbol.isRequired
};

export default DeleteButton;
