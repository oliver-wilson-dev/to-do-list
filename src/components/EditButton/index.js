import React from 'react';
import PropTypes from 'prop-types';
import Button from '../Button';

const EditButton = ({
  setEditState, id, editing, description
}) => {
  const onClick = () => setEditState({ id, editState: !editing, description });

  return (
    <Button onClick={onClick}>
      {editing
        ? <span role="img" aria-label="apply changes">✅</span>
        : <span role="img" aria-label="edit this task">✏️</span>
      }
    </Button>
  );
};

EditButton.propTypes = {
  setEditState: PropTypes.func.isRequired,
  id: PropTypes.symbol.isRequired,
  editing: PropTypes.bool.isRequired,
  description: PropTypes.string.isRequired
};

export default EditButton;
