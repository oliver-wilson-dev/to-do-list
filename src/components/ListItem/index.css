@import "../../styles/variables.css";

.listItem {
    margin: 10px 10px 10px 0px;
    list-style-type: none;
    text-align: left;
    display: flex;
    justify-content: space-between;
    align-items: center;
    box-shadow: 4px 4px 0px var(--red);
    border: 3px solid var(--primaryTextColor);
    border-radius: inherit;
    padding: 5px;
    font-size: 14px;
    flex-shrink: 0;

    &:last-of-type {
        padding-bottom: 0px;
    }
}

.description {
    margin-right: auto;
    display: inline-block;
    flex-basis: 60%;
}

.listItemNth0 {
    box-shadow: 4px 4px 0px var(--red);
}

.listItemNth1 {
    box-shadow: 4px 4px 0px var(--green);
}

.listItemNth2 {
    box-shadow: 4px 4px 0px var(--blue);
}

.listItemNth3 {
    box-shadow: 4px 4px 0px var(--yellow);
}

.listItemNth4 {
    box-shadow: 4px 4px 0px var(--purple);
} 

.textArea {
    flex-basis: 60%;
    height: 65px;
}

.complete {
    text-decoration: line-through;
}

.actions {
    display: flex;
    align-items: center;
    flex-direction: column;
    flex-basis: 20%;

    & > * {
        margin: 3px 0px;
    }
}