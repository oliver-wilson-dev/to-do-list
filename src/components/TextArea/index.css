.textArea {
    resize: none;
    overflow: auto;
    background-color: var(--primaryBackgroundColour);
    font-size: inherit;
    padding: 5px 10px;
    border: 3px solid var(--primaryTextColor);
    margin-right: 5px;
}