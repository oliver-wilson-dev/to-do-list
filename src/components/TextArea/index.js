import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';

const TextArea = ({
  onChange, value, additionalStyles, reference
}) => {
  React.useEffect(() => {
    reference.current.focus();
  }, []);

  return (
    <textarea
      className={`${styles.textArea} ${additionalStyles || ''}`}
      placeholder="Start typing..."
      maxLength={50}
      onChange={onChange}
      value={value}
      ref={reference}
    />
  );
};

TextArea.propTypes = {
  additionalStyles: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  reference: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any })
  ]).isRequired
};

export default TextArea;
