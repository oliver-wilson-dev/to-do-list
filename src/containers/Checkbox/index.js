import { connect } from 'react-redux';
import Checkbox from '../../components/Checkbox';
import markComplete from '../../state/actions/markComplete';

const mapDispatchToProps = dispatch => ({
  markComplete:
  ({ id }) => dispatch(markComplete({ id }))
});

export default connect(undefined, mapDispatchToProps)(Checkbox);
