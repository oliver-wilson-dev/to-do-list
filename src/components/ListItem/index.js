import React from 'react';
import PropTypes from 'prop-types';
import styles from './index.css';
import DeleteButton from '../../containers/DeleteButton';
import EditButton from '../../containers/EditButton';
import Checkbox from '../../containers/Checkbox';
import TextArea from '../TextArea';

const ListItem = ({
  description, id, index, editing, complete
}) => {
  const [newDescription, setDescription] = React.useState(description);

  const onChange = ({ target: { value: newDescription } }) => {
    setDescription(newDescription);
  };

  const ref = React.useRef(null);

  return (
    <li className={`${styles.listItem} ${styles[`listItemNth${index % 5}`]}`}>
      {editing
        ? (
          <TextArea
            onChange={onChange}
            value={newDescription}
            reference={ref}
            additionalStyles={styles.textArea}
          />
        )
        : <span className={`${styles.description} ${complete ? styles.complete : ''}`}>{description}</span>}
      <div className={styles.actions}>
        <Checkbox id={id} complete={complete} />
        <EditButton id={id} description={newDescription} editing={editing} />
        <DeleteButton id={id} />
      </div>
    </li>
  );
};

ListItem.propTypes = {
  description: PropTypes.string.isRequired,
  id: PropTypes.symbol.isRequired,
  index: PropTypes.number.isRequired,
  editing: PropTypes.bool.isRequired,
  complete: PropTypes.bool.isRequired
};

export default ListItem;
