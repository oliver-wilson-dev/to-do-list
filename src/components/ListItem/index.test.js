import React from 'react';
import { shallow } from 'enzyme';
import ListItem from './index';

import DeleteButton from '../../containers/DeleteButton';
import EditButton from '../../containers/EditButton';
import Checkbox from '../../containers/Checkbox';
import TextArea from '../TextArea';

jest.mock('../Button', () => {
  const Button = () => null;

  return Button;
});

jest.mock('../../containers/DeleteButton', () => {
  const DeleteButton = () => null;

  return DeleteButton;
});

jest.mock('../../containers/EditButton', () => {
  const EditButton = () => null;

  return EditButton;
});

jest.mock('../../containers/Checkbox', () => {
  const Checkbox = () => null;

  return Checkbox;
});

jest.mock('../TextArea', () => {
  const TextArea = () => null;

  return TextArea;
});

const mockIndex = 0;
const mockId = Symbol('test-id');
const mockDescription = 'test-description';
const mockComplete = false;
const mockEditing = false;

const defaultProps = {
  description: mockDescription,
  id: mockId,
  index: mockIndex,
  editing: mockEditing,
  complete: mockComplete
};

const render = overrideProps => shallow(<ListItem {...defaultProps} {...overrideProps} />);

describe('<ListItem/> component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should exist', () => {
    expect(render().exists()).toBe(true);
  });

  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  test.each`
  index| expected    
  ${0} | ${0} 
  ${1} | ${1} 
  ${2} | ${2}
  ${3} | ${3}
  ${4} | ${4}
  ${5} | ${0}
  ${6} | ${1}
  ${7} | ${2}
  ${8} | ${3}
  ${9} | ${4}
`('when the index is $index it sets the correct className based on the index of the task', ({ index, expected }) => {
  expect(render({ index }).find('li').prop('className')).toMatch(`listItemNth${expected}`);
});

  describe('when the task is being edited', () => {
    it('should render correctly', () => {
      expect(render({ editing: true })).toMatchSnapshot();
    });

    describe('and the editing is done', () => {
      it('should set the description to be the new description provided', () => {
        const mockNewDescription = 'test-new-description';

        const wrapper = render({ editing: true });
        wrapper.find(TextArea).simulate('change', { target: { value: mockNewDescription } });
        expect(wrapper.find(EditButton).prop('description')).toEqual(mockNewDescription);
      });
    });
  });

  describe('when the task is not being edited', () => {
    describe('and the task is complete', () => {
      it('should render correctly', () => {
        expect(render({ editing: false, complete: true })).toMatchSnapshot();
      });
    });

    describe('and the task is not complete', () => {
      it('should render correctly', () => {
        expect(render({ editing: false, complete: false })).toMatchSnapshot();
      });
    });
  });
});
