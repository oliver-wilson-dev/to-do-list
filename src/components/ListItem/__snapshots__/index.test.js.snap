// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem/> component should render correctly 1`] = `
<li
  className="listItem listItemNth0"
>
  <span
    className="description "
  >
    test-description
  </span>
  <div
    className="actions"
  >
    <Checkbox
      complete={false}
      id={Symbol(test-id)}
    />
    <EditButton
      description="test-description"
      editing={false}
      id={Symbol(test-id)}
    />
    <DeleteButton
      id={Symbol(test-id)}
    />
  </div>
</li>
`;

exports[`<ListItem/> component when the task is being edited should render correctly 1`] = `
<li
  className="listItem listItemNth0"
>
  <TextArea
    additionalStyles="textArea"
    onChange={[Function]}
    reference={Symbol(test-ref)}
    value="test-description"
  />
  <div
    className="actions"
  >
    <Checkbox
      complete={false}
      id={Symbol(test-id)}
    />
    <EditButton
      description="test-description"
      editing={true}
      id={Symbol(test-id)}
    />
    <DeleteButton
      id={Symbol(test-id)}
    />
  </div>
</li>
`;

exports[`<ListItem/> component when the task is not being edited and the task is complete should render correctly 1`] = `
<li
  className="listItem listItemNth0"
>
  <span
    className="description complete"
  >
    test-description
  </span>
  <div
    className="actions"
  >
    <Checkbox
      complete={true}
      id={Symbol(test-id)}
    />
    <EditButton
      description="test-description"
      editing={false}
      id={Symbol(test-id)}
    />
    <DeleteButton
      id={Symbol(test-id)}
    />
  </div>
</li>
`;

exports[`<ListItem/> component when the task is not being edited and the task is not complete should render correctly 1`] = `
<li
  className="listItem listItemNth0"
>
  <span
    className="description "
  >
    test-description
  </span>
  <div
    className="actions"
  >
    <Checkbox
      complete={false}
      id={Symbol(test-id)}
    />
    <EditButton
      description="test-description"
      editing={false}
      id={Symbol(test-id)}
    />
    <DeleteButton
      id={Symbol(test-id)}
    />
  </div>
</li>
`;
